src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java

```diff
@@ -10,9 +11,28 @@
 public class OnHitEvent {
 
     @SubscribeEvent
-    public static void onHit(LivingIncomingDamageEvent event) {
-        float dmg = OnHitEffect.critical(event.getAmount(), (LivingEntity) event.getSource().getDirectEntity(),
-                event.getEntity(), event.getEntity().getRandom());
-        event.setAmount(dmg);
+    public static void onCriticalHit(LivingIncomingDamageEvent event) {
+        if (event.getSource().getDirectEntity() instanceof LivingEntity livingAttacker) {
+            float dmg = OnHitEffect.critical(event.getAmount(), livingAttacker, event.getEntity(),
+                    event.getEntity().getRandom());
+            event.setAmount(dmg);
+        }
+    }
+
+    @SubscribeEvent
+    public static void onThornsProc(LivingIncomingDamageEvent event) {
```

Comment:
Would suggest method name would be better as `checkThornsProc` (it isn't called when thorns are triggered)